linetools/lists/linelist.py

```diff
@@ -0,0 +1,236 @@
+
+from linetools.lists import parse as lilp
+
+#
+class LineList(object):
+    '''Class to over-load Spectrum1D for new functionality not yet in specutils
+
+    Parameters:
+    ----------
+    llst_keys: str or list  
+      Input to grab line list.  Current options are:
+       'ISM'     :: "All" ISM lines
+       'Strong'  :: Strong ISM lines
+       'HI'      :: HI Lyman series
+       'H2'      :: H2 (Lyman-Werner)
+       'CO'      :: CO UV band-heads [not yet implemented]
+       'EUV'     :: Key EUV lines (for CASBAH project)
+       'Gal_E'   :: Galaxy emission lines (HII)
+       'Gal_A'   :: Galaxy absorption lines (stellar)
+       'AGN'     :: Key AGN lines
+    gd_lines: list, optional
+      List of wrest for lines to use (drawn from input linelist)
+    verbose: bool, optional
+      Give info galore if True
+    '''
+    # Init
+    def __init__(self, llst_keys, gd_lines=None, verbose=True):
+
+        # Error catching
+        if type(llst_keys) not in [str,list,unicode]:
+            raise TypeError('LineList__init__: Wrong type for LineList input')
+
+        # Save
+        if type(llst_keys) in [str,unicode]:
+            self.lists = [llst_keys]
+        else:
+            self.lists = llst_keys
+
+        # Load Data
+        self.load_data()
+
+        # Set lines for use
+        self.set_lines(gd_lines=gd_lines, verbose=verbose)
+
+    # 
+    def load_data(self, tol=1e-3*u.AA):
+        ''' Grab the data for the lines of interest
+        '''
+        # Import
+        reload(lilp)
+
+        # Define datasets: In order of Priority
+        dataset = {
+            'ism': [lilp.parse_morton03,lilp.read_verner94], # Verner 1994, Morton 2003 
+            'molecules': [lilp.read_H2]   # H2 
+            }
+
+        # Loop on lists
+        sets = []
+        for llist in self.lists:
+            if str(llist) == 'H2':
+                sets.append('molecules')
+            elif str(llist) == 'ISM':
+                sets.append('ism')
+            elif str(llist) == 'Strong':
+                sets.append('ism')
+            else:
+                import pdb
+                pdb.set_trace()
+                raise ValueError('load_data: Not ready for this: {:s}'.format(llist))
+
+        full_table = None
+        all_func = []
+        # Loop on data sets
+        for iset in sets:
+            # Loop on data sources
+            for func in dataset[iset]:
+                # Query if read already
+                if func not in all_func:
+                    # Read
+                    table = func()
+                    if full_table is None:
+                        full_table = table
+                    else:
+                        # Unique values
+                        wrest = full_table['wrest']
+                        newi = []
+                        for jj,row in enumerate(QTable(table)): # QTable for units
+                            mt = np.where(np.abs(row['wrest']-wrest) < tol)[0]
```

Comment:
Well, that is mainly what the tolerance is for.
I don’t know of any that are the same at 1e-3 and
we could push it to 1e-4.  

But instead of name it would be fast to do Z+ion.

Your call,

X


On Jun 21, 2015, at 1:12 PM, Neil Crighton <notifications@github.com> wrote:

> In linetools/lists/linelist.py:
> 
> > +        # Loop on data sets
> > +        for iset in sets:
> > +            # Loop on data sources
> > +            for func in dataset[iset]:
> > +                # Query if read already
> > +                if func not in all_func:
> > +                    # Read
> > +                    table = func()
> > +                    if full_table is None:
> > +                        full_table = table
> > +                    else:
> > +                        # Unique values
> > +                        wrest = full_table['wrest']
> > +                        newi = []
> > +                        for jj,row in enumerate(QTable(table)): # QTable for units
> > +                            mt = np.where(np.abs(row['wrest']-wrest) < tol)[0]
> Should we also check the transition/atom name here? i.e. do we need to worry about transitions from different species with almost identical rest wavelengths?
> 
> —
> Reply to this email directly or view it on GitHub.
> 
> !DSPAM:826,55871cd71841690719359!

----------------------------------------------
J. Xavier Prochaska
UCO/Lick Observatory
1156 High St.
UC Santa Cruz
Santa Cruz, CA 95064

http://www.ucolick.org/~xavier/
http://imps.ucolick.org/
http://www.qpqsurvey.org/
http://k1dm3.ucolick.org/
http://www.ucolick.org/~xavier/IDL/index.html
https://github.com/profxj/xastropy

831-459-2135 (Direct)
831-459-2991 (UCO/Lick Main)
831-459-5244 (Fax)